CppImport/src/macro/ClangHelper.cpp

```diff
@@ -0,0 +1,126 @@
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "ClangHelper.h"
+
+namespace CppImport {
+
+void ClangHelper::setSourceManager(const clang::SourceManager* sourceManager)
+{
+	sourceManager_ = sourceManager;
+}
+
+void ClangHelper::setPreprocessor(const clang::Preprocessor* preprocessor)
+{
+	preprocessor_ = preprocessor;
+}
+
+QString ClangHelper::spelling(clang::SourceLocation loc)
+{
+	return spelling(loc, loc);
+}
+
+QString ClangHelper::spelling(clang::SourceRange range)
+{
+	return spelling(range.getBegin(), range.getEnd());
+}
+
+QString ClangHelper::spelling(clang::SourceLocation start, clang::SourceLocation end)
+{
+	clang::SourceLocation b = sourceManager_->getSpellingLoc(start);
+	clang::SourceLocation e = clang::Lexer::getLocForEndOfToken(sourceManager_->getSpellingLoc(end), 0, *sourceManager_,
+																					preprocessor_->getLangOpts());
+
+	auto length = sourceManager_->getCharacterData(e) - sourceManager_->getCharacterData(b);
+	if (length > 1000000) return "ERROR_IN_GET_SPELLING";
+
+	try
+	{
+		return 0 < length ? QString::fromStdString(std::string(sourceManager_->getCharacterData(b), length)) : "";
+	}
+	catch (...)
+	{
+		return "ERROR_IN_GET_SPELLING";
+	}
+}
+
+clang::SourceLocation ClangHelper::immediateMacroLoc(clang::SourceLocation Loc)
+{
+	if (Loc.isMacroID())
+	{
+		while (1)
+		{
+			auto FID = sourceManager_->getFileID(Loc);
+			const clang::SrcMgr::SLocEntry *E = &sourceManager_->getSLocEntry(FID);
+			if (!E->isExpansion())
+				break;
+			const clang::SrcMgr::ExpansionInfo &Expansion = E->getExpansion();
+			Loc = Expansion.getExpansionLocStart();
+			if (!Expansion.isMacroArgExpansion())
+				break;
+
+			Loc = sourceManager_->getImmediateExpansionRange(Loc).first;
+			auto SpellLoc = Expansion.getSpellingLoc();
+			if (SpellLoc.isFileID())
+				break;
+
+			auto MacroFID = sourceManager_->getFileID(Loc);
+			if (sourceManager_->isInFileID(SpellLoc, MacroFID))
+				break;
+
+			Loc = SpellLoc;
```

Comment:
yes. already the first break condition is eventually going to be false (after the top level macro call). In fact most calls are expected not to iterate multiple times.